### ISSUE TITLE
Model Performance

### ISSUE THREAD

**Issue by Walden-ea** [AI-Application-and-Integration-Lab/DGUA_FAS#1]
I conducted several experiments with the model: both on my dataset and on a portion of CelebA-Spoof. In any scenario, the model seemed to produce a huge amount of false positives: 60% of real images were marked as spoofed.

So, there are some questions:
1. What kind of image preprocessing do you use? I see torch normalization transform, but do you do any preprocessing before that? To be more specific:
	1. Do you crop the images using the bounding boxes? 
	2. Do you make the bounding boxes square or just resize the cropped images to 224x224? 
	3. Do you do any face alignment?
2. Is the model presented in the repository robust enough for regular use?

Thank you!

**Comment by zero-suger** [comment]
> I conducted several experiments with the model: both on my dataset and on a portion of CelebA-Spoof. n any scenario, the model seemed to produce a huge amount of false positives: 60% of real images were marked as spoofed.
> 
> So, there are some questions:
> 
> 1. What kind of image preprocessing do you use? I see torch normalization transform, but do you do any preprocessing before that? To be more specific:
>    
>    1. Do you crop the images using the bounding boxes?
>    2. Do you make the bounding boxes square or just resize the cropped images to 224x224?
>    3. Do you do any face alignment?
> 2. Is the model presented in the repository robust enough for regular use?
> 
> Thank you!

Hi, I am also researcher like you and I have been training and testing this model with big 9 datasets (most are private so I can't tell the names of datasets). And I got impressive results on testing on new images with IP camera. Answer to your questions: 

1. All data pre processing is based on SSDG(https://github.com/taylover-pei/SSDG-CVPR2020), please read, all paper carefully and check SSDG github and back to DGUA-FAS, hopefully you can understand easily. 

 Main point -> detect face (RetinaFace, MTCNN, FACEB, CphereFace, etc) -> crop and save  
 
 i. It is your choice bbox or other. I just used third party RetinaFace installation and  detect face and cropped faces with +30px 
    margin around face.
 ii. Cropped image sizes are 256 X 256. Just all.
 iii. please check ml-cvnets -> cvnets -> models -> classification -> movbilevit.py and ROOT -> configs -> mobilevit_s.yaml
 2. I am not sure yet, but now I have trained model with max 5000 iterations, and in C - O - M - CelebFAS, I got more than 
 98.3%. (I am keep testing on it.)
 
 Please read : https://arxiv.org/pdf/2310.11758.pdf (paper and back on your code and check your code from beggining and I hope you will get impressive results (: )

**Comment by rlczddl** [comment]
Hi, Do you understand the line "loss = criterion(avg_single_video_output, torch.where(avg_single_video_target == 1, 1, 0))
        acc_valid = accuracy(avg_single_video_output, torch.where(avg_single_video_target == 1, 1, 0), topk=(1,))" in eval.py? The  
avg_single_video_output's size is b*3, which label is 0, 1, 2, but the label of avg_single_video_target  after torch.where is 0, 1, how to  compare?

**Comment by zero-suger** [comment]
> Hi, Do you understand the line "loss = criterion(avg_single_video_output, torch.where(avg_single_video_target == 1, 1, 0)) acc_valid = accuracy(avg_single_video_output, torch.where(avg_single_video_target == 1, 1, 0), topk=(1,))" in eval.py? The avg_single_video_output's size is b*3, which label is 0, 1, 2, but the label of avg_single_video_target after torch.where is 0, 1, how to compare?
hi, 
eval.py? which file you mean ?

**Comment by rlczddl** [comment]
Sorry, it is in /util/evaluate.py.

**Comment by zero-suger** [comment]
> Sorry, it is in /util/evaluate.py.

Actually, I didn't apply VideoID and loss calculation as it was in the code. I just did binary classification with images and don't consider videoID. Instead of this, i did mannually just certain (random) videoIDs  and load images and train and save map scores of 2nd class to .txt file and calculate EER with ROC. Because my dataset was different and not ID based. Sorry ~

**Comment by rlczddl** [comment]
Thanks for the reply, and I use it for two categories, too. I'm just curious about the code the author posted here and I can't understand it.

**Comment by stonejane** [comment]
> > I conducted several experiments with the model: both on my dataset and on a portion of CelebA-Spoof. n any scenario, the model seemed to produce a huge amount of false positives: 60% of real images were marked as spoofed.
> > So, there are some questions:
> > 
> > 1. What kind of image preprocessing do you use? I see torch normalization transform, but do you do any preprocessing before that? To be more specific:
> >    
> >    1. Do you crop the images using the bounding boxes?
> >    2. Do you make the bounding boxes square or just resize the cropped images to 224x224?
> >    3. Do you do any face alignment?
> > 2. Is the model presented in the repository robust enough for regular use?
> > 
> > Thank you!
> 
> Hi, I am also researcher like you and I have been training and testing this model with big 9 datasets (most are private so I can't tell the names of datasets). And I got impressive results on testing on new images with IP camera. Answer to your questions:
> 
>     1. All data pre processing is based on SSDG(https://github.com/taylover-pei/SSDG-CVPR2020), please read, all paper carefully and check SSDG github and back to DGUA-FAS, hopefully you can understand easily.
> 
> 
> Main point -> detect face (RetinaFace, MTCNN, FACEB, CphereFace, etc) -> crop and save
> 
> i. It is your choice bbox or other. I just used third party RetinaFace installation and detect face and cropped faces with +30px margin around face. ii. Cropped image sizes are 256 X 256. Just all. iii. please check ml-cvnets -> cvnets -> models -> classification -> movbilevit.py and ROOT -> configs -> mobilevit_s.yaml 2. I am not sure yet, but now I have trained model with max 5000 iterations, and in C - O - M - CelebFAS, I got more than 98.3%. (I am keep testing on it.)
> 
> Please read : https://arxiv.org/pdf/2310.11758.pdf (paper and back on your code and check your code from beggining and I hope you will get impressive results (: )

I got a problem in line 14 of 'base_cls.py' applied by the author('from ...misc.profiler import module_profile'), and I could not find 'profiler' in 'cvnets/misc'. Could u kindly tell me how to solve this problem? Thx a lot!!!🙇‍♀️ @zero-suger

**Comment by zero-suger** [comment]
Good Morning ~ 

1) create a conda environment 
2) git clone https://github.com/AI-Application-and-Integration-Lab/DGUA_FAS.git
3) cd AI-Application-and-Integration-Lab
4) git clone https://github.com/apple/ml-cvnets
5) cd ml-cvnets 
6) git checkout 84d992f413e52c0468f86d23196efd9dad885e6f
7) go to ml-cvnets/cvnets/models/classification and find base_cls.py file and delete and https://drive.google.com/file/d/1shq23SpC4X2OoYFELFjHMWpYyolmMEpj/view pase this file as a base_cls.py
8) pip install -r requirements.txt
9) pip install --editable .
10) pip install pandas 
11) pip install tensorboard
12) data preparation and start training 
Note : make sure you do all steps under one environment (conda or etc.) when you change the conda environment or your working environment you have to do '6) git checkout 84d992f413e52c0468f86d23196efd9dad885e6f' (more like initialization ) and this one may be reason in your case and files can't find 

good luck (:

**Comment by stonejane** [comment]
> Good Morning ~
> 
>     1. create a conda environment
> 
>     2. git clone https://github.com/AI-Application-and-Integration-Lab/DGUA_FAS.git
> 
>     3. cd AI-Application-and-Integration-Lab
> 
>     4. git clone https://github.com/apple/ml-cvnets
> 
>     5. cd ml-cvnets
> 
>     6. git checkout 84d992f413e52c0468f86d23196efd9dad885e6f
> 
>     7. go to ml-cvnets/cvnets/models/classification and find base_cls.py file and delete and https://drive.google.com/file/d/1shq23SpC4X2OoYFELFjHMWpYyolmMEpj/view pase this file as a base_cls.py
> 
>     8. pip install -r requirements.txt
> 
>     9. pip install --editable .
> 
>     10. pip install pandas
> 
>     11. pip install tensorboard
> 
>     12. data preparation and start training
>         Note : make sure you do all steps under one environment (conda or etc.) when you change the conda environment or your working environment you have to do '6) git checkout 84d992f413e52c0468f86d23196efd9dad885e6f' (more like initialization ) and this one may be reason in your case and files can't find
> 
> 
> good luck (:

wow!!It worked!Thanks for ur great help!🙇‍♀️

**Comment by p458732** [comment]
> Thanks for the reply, and I use it for two categories, too. I'm just curious about the code the author posted here and I can't understand it.

The number of classes is set to three because we adopted the DiVT approach for classifying attack types (real, print, replay) instead of focusing on the distinction between real and spoofing.
